### PR TITLE
REPO-5571 : Change the tempFileCleanerTrigger to work at scale

### PR DESCRIPTION
**Pull request by tzclucian**
- added nr of files and time limits

### CODE REVIEW

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 1]
```diff
@@ -468,7 +502,17 @@ private static int removeFiles(File directory, long removeBefore, long longLifeB
                         {
                             logger.debug("Deleting temp file: " + file);
                         }
-                        file.delete();
+                        // only delete if the limits allow
+                        if (maxFilesToDelete != null && maxFilesToDelete.get() <= 0 ||
+                            maxTimeToRun != null && ((jobStartTime + maxTimeToRun.toMillis()) < System.currentTimeMillis()))
+                        {
+                            return 0;
```
Comment (cristianturlica) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:509]: This should be `return count`.

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 2]
```diff
@@ -376,16 +389,37 @@ public void execute(JobExecutionContext context) throws JobExecutionException
...
+            try
+            {
+                final String strMaxFilesToDelete = (String) context.getJobDetail().getJobDataMap().get(KEY_MAX_FILES_TO_DELETE);
+                maxFilesToDelete = new AtomicLong(Long.parseLong(strMaxFilesToDelete));
+            }
+            catch (Exception e)
+            {
+                logger.warn(e);
+                throw new JobExecutionException("Invalid job data " + KEY_MAX_FILES_TO_DELETE + ": " + strProtectHours);
+            }
```
Comment (cristianturlica) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:402]: This could be refactored to a method as it is repeated just below.

**Comment by tzclucian** [comment]
For testin purposes I would suggest a configuration similar to this:

```
   <bean id="tempFileCleanerTrigger" class="org.springframework.scheduling.quartz.CronTriggerFactoryBean">
        <property name="cronExpression" value="0 0/5 * * * ?" /> <!-- Repeat hourly on the half hour -->
        <property name="startDelay" value="${system.cronJob.startDelayMilliseconds}"/>
        <property name="jobDetail">
            <bean class="org.springframework.scheduling.quartz.JobDetailFactoryBean">
                <property name="jobClass" value="org.alfresco.util.TempFileProvider$TempFileCleanerJob"/>
                <property name="jobDataAsMap">
                    <map>
                        <entry key="protectHours" value="0"/>
                        <entry key="maxFilesToDelete" value="${system.tempFileCleaner.maxFilesToDelete:#{null}}"/>
                        <entry key="maxTimeToRun" value="${system.tempFileCleaner.maxTimeToRun:#{null}}"/>
                    </map>
                </property>
            </bean>
        </property>
    </bean>
    <bean id="webscripts.tempFileCleanerTrigger" class="org.springframework.scheduling.quartz.CronTriggerFactoryBean">
        <property name="cronExpression" value="0 0/5 * * * ?" /> <!-- Repeat hourly on the start hour -->
        <property name="startDelay" value="${system.cronJob.startDelayMilliseconds}"/>
        <property name="jobDetail">
            <bean class="org.springframework.scheduling.quartz.JobDetailFactoryBean">
                <property name="jobClass" value="org.alfresco.util.TempFileProvider$TempFileCleanerJob"/>
                <property name="jobDataAsMap">
                    <map>
                        <entry key="protectHours" value="0"/>
                        <entry key="directoryName" value="${webscripts.tempDirectoryName}"/>
                        <entry key="maxFilesToDelete" value="${system.tempFileCleaner.maxFilesToDelete:#{null}}"/>
                        <entry key="maxTimeToRun" value="${system.tempFileCleaner.maxTimeToRun:#{null}}"/>
                    </map>
                </property>
            </bean>
        </property>
    </bean>
```

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 3]
```diff
@@ -376,16 +389,47 @@ public void execute(JobExecutionContext context) throws JobExecutionException
...
+                {
+                    final String strMaxFilesToDelete = (String) oMaxFilesToDelete;
+                    maxFilesToDelete = new AtomicLong(Long.parseLong(strMaxFilesToDelete));
+                    logger.debug("Set the maximum number of temp files to be deleted to: " + maxFilesToDelete.get());
+                }
+            }
+            catch (Exception e)
+            {
+                logger.warn(e);
+                throw new JobExecutionException("Invalid job data, maxFilesToDelete: " + KEY_MAX_FILES_TO_DELETE + ": " + strProtectHours);
```
Comment (DenisGabriela) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:406]: This message does not seem correct.

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 4]
```diff
@@ -468,7 +513,17 @@ private static int removeFiles(File directory, long removeBefore, long longLifeB
...
+                        // only delete if the limits allow
+                        if (maxFilesToDelete != null && maxFilesToDelete.get() <= 0 ||
+                            maxTimeToRun != null && ((jobStartTime + maxTimeToRun.toMillis()) < System.currentTimeMillis()))
+                        {
+                            return count;
+                        }
+                        else
+                        {
+                            file.delete();
+                            maxFilesToDelete.decrementAndGet();
```
Comment (DenisGabriela) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:525]: Possible NPE?

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 5]
```diff
@@ -376,16 +389,47 @@ public void execute(JobExecutionContext context) throws JobExecutionException
             }
 
             String directoryName = (String) context.getJobDetail().getJobDataMap().get(KEY_DIRECTORY_NAME);
+
+            try
+            {
+                final Object oMaxFilesToDelete = context.getJobDetail().getJobDataMap().get(KEY_MAX_FILES_TO_DELETE);
```
Comment (DenisGabriela) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:395]: Relates to the comment below - what if `maxFilesToDelete` is not set?

File: core/src/main/java/org/alfresco/util/TempFileProvider.java (hunk 4 above)
Comment (tzclucian) [reply core/src/main/java/org/alfresco/util/TempFileProvider.java:525]: Thanks, updated.

**Review by DenisGabriela — APPROVED** [review]

File: core/src/main/java/org/alfresco/util/TempFileProvider.java [hunk 6]
```diff
@@ -468,7 +521,32 @@ private static int removeFiles(File directory, long removeBefore, long longLifeB
...
                             logger.debug("Deleting temp file: " + file);
                         }
-                        file.delete();
+                        // only delete if the limits allow
+                        if (maxFilesToDelete != null && maxFilesToDelete.get() <= 0 ||
+                            maxTimeToRun != null && ((jobStartTime + maxTimeToRun.toMillis()) < System.currentTimeMillis()))
+                        {
+                            return count;
+                        }
+                        else
```
Comment (CezarLeahu) [inline core/src/main/java/org/alfresco/util/TempFileProvider.java:530]: No need to use `else` after a `if { return/break/continue }` statement. It's redundand and needlessly  expands the code horizontally.

**Review by CezarLeahu — APPROVED** [review]

**Review by TheOutsSeeker — APPROVED** [review]
LGTM

**Pull request merged**